src/3.md

```diff
@@ -56,32 +56,55 @@ match f {
 }
 ```
 
-We could also use `unwrap`, which panics if the `Result` is an `Err(E)`, and returns `T` if not. `expect` is the same, except we can specify an error message
+We could also use `unwrap`, which panics if the `Result` is an `Err(E)`, and returns `T` if not. `expect` behaves the same, except we can specify an error message
 
 ```rust, editable
 use std::{fs, io};
-let f: Result<fs::File, io::Error> = fs::File::open("foo.txt");
+fn main() {
+    let f: Result<fs::File, io::Error> = fs::File::open("foo.txt");
+    let file = f.expect("Could not open file");
+}
+```
 
-let file = f.expect("Could not open file");
+We could also propagate the `Result` error back up to the caller. If we wanted to do this manually, we would need something like:
+
+```rust, ignore
+fn maybe_func() -> Result<T, E>;
+
+let x = match maybe_func() {
+  Ok(x) => x,
+  Err(e) => return Err(e),
+};
+// equivalent to
+let x = maybe_func()?;
 ```
 
-We could also propagate the error back up to the caller. In functions that return a `Result<T, E>` or `Option<T>`, we can use the `?` operator to bubble the error back up, where the return type is compatible.
+In functions that return a `Result<T, E>` or `Option<T>`, we can use the `?` operator as a shorthand to bubble the error back up, where the return type is compatible. 
 
 ```rust, editable
 use std::{fs, io};
 fn open_my_file() -> Result<String, io::Error>  {
     let contents: String = fs::read_to_string("foo.txt")?;
     Ok(contents)
 }
+
+fn main() {
+    match open_my_file() {
+        Ok(text) => println!("File contents {text}"),
+        Err(e) => println!("Could not read file"),
+    }
+}
```

Comment:
you can also bubble errors up out of `main`, as the return type can be `Result<(), Box<dyn std::error:Error>`